File: cses/math/exponentiation_2/exp2.zig
const std = @import("std");
const io = std.io;
const math = std.math;

const MOD: i64 = 1000000007;

fn power(a: i64, b: i64, p: i64) i64 {
    var res: i64 = 1;
    var a_var = a;
    var b_var = b;
    while (b_var > 0) : (b_var >>= 1) {
        if (b_var & 1 != 0) {
            res = uncheckedMulMod(res, a_var, p);
        }
        a_var = uncheckedMulMod(a_var, a_var, p);
    }
    return res;
}

fn uncheckedMulMod(a: i64, b: i64, p: i64) i64 {
    return math.mod(i64, a * b, p);
}

pub fn main() !void {
    const stdout = io.getStdOut().writer();
    var stdin = io.getStdIn().reader();
    var buffer: [100]u8 = undefined;

    // Read the first line for the number of test cases
    const line = try stdin.readUntilDelimiterOrEof(&buffer, '\n');
    if (line) |l| {
        const n: usize = std.fmt.parseInt(usize, l, 10) catch 0;

        // Loop through each test case
        for (std.mem.repeat(u8, 0, n)) |_| {
            const line = try stdin.readUntilDelimiterOrEof(&buffer, '\n');
            if (line) |l| {
                var iter = std.mem.split(l, " ");
                const a = std.fmt.parseInt(i64, iter.next().?, 10) catch 0;
                const b = std.fmt.parseInt(i64, iter.next().?, 10) catch 0;
                const c = std.fmt.parseInt(i64, iter.next().?, 10) catch 0;

                const ans = power(a, power(b, c, MOD - 1), MOD);
                try stdout.print("{}\n", .{ans});
            }
        }
    }
}

File: cses/math/exponentiation_2/exp2.rs
use std::io::{self, BufRead, BufReader, Write, BufWriter};
 
const MOD: i64 = 1000000007;
const PHI: i64 = MOD - 1; // Euler's totient function value for MOD
 
#[inline(always)]
fn power(mut a: i64, mut b: i64, p: i64) -> i64 {
    let mut result: i64 = 1;
    while b > 0 {
        if b & 1 != 0 {
            result = (result * a) % p;
        }
        a = (a * a) % p;
        b >>= 1;
    }
    result
}
 
fn main() {
    let stdin = io::stdin();
    let stdout = io::stdout();
    let reader = BufReader::new(stdin.lock());
    let mut writer = BufWriter::new(stdout.lock());
 
    let mut lines = reader.lines();
    let n: usize = lines.next().unwrap().unwrap().trim().parse().unwrap();
 
    for line in lines.take(n) {
        let line = line.unwrap();
        let mut nums = [0i64; 3];
        let mut idx = 0;
        let mut current_num = 0i64;
 
        for byte in line.bytes() {
            if byte.is_ascii_digit() {
                current_num = 10 * current_num + (byte - b'0') as i64;
            } else if byte == b' ' || byte == b'\n' {
                nums[idx] = current_num;
                idx += 1;
                current_num = 0;
            }
        }
        nums[idx] = current_num;
 
        let ans = power(nums[0], power(nums[1], nums[2], PHI), MOD);
        writeln!(writer, "{}", ans).unwrap();
    }
}
